### PR TITLE
[Serve] Fix ServeHandle serialization

### LINKED ISSUE #13180: [Serve] ServeHandle no longer serializable
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
```
from ray import serve
import ray

def f(_): pass

client = serve.start()
client.create_backend("f", f)
client.create_endpoint("f", backend="f")
handle = client.get_handle("f")

ray.put(handle)
```

ServeHandle can no longer be passed around to other Ray tasks and the error is really hard to debug

```
2021-01-04 21:15:59,360	INFO services.py:1173 -- View the Ray dashboard at http://127.0.0.1:8265
(pid=6369) 2021-01-04 21:16:01,432	INFO controller.py:346 -- Starting router with name 'hRhwaS:SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:192.168.31.141-0' on node 'node:192.168.31.141-0' listening on '127.0.0.1:8000'
(pid=6366) INFO:     Started server process [6366]
Traceback (most recent call last):
  File "a.py", line 11, in <module>
    ray.put(handle)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/worker.py", line 1411, in put
    object_ref = worker.put_object(value, pin_object=True)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/worker.py", line 262, in put_object
    serialized_value = self.get_serialization_context().serialize(value)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/serialization.py", line 406, in serialize
    return self._serialize_to_msgpack(value)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/serialization.py", line 386, in _serialize_to_msgpack
    self._serialize_to_pickle5(metadata, python_objects)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/serialization.py", line 346, in _serialize_to_pickle5
    raise e
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/serialization.py", line 343, in _serialize_to_pickle5
    value, protocol=5, buffer_callback=writer.buffer_callback)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/Users/simonmo/miniconda3/envs/py37/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 563, in dump
    return Pickler.dump(self, obj)
  File "stringsource", line 2, in ray._raylet.Count.__reduce_cython__
TypeError: self.c_tag_keys,self.metric cannot be converted to a Python object for pickling
(pid=6369) 2021-01-04 21:16:03,924	INFO controller.py:753 -- Registering route 'f' to endpoint 'f' with methods '['GET']'.
2021-01-04 21:16:06,021	INFO api.py:65 -- Shutting down Ray Serve because client went out of scope. To prevent this, either keep a reference to the client object or use serve.start(detached=True).
```

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds `__reduce__` methods to ServeHandle and `ThreadProxiedRouter` to fix serialization of ServeHandles.  Upon deserialization, the constructor is simply called again, and a new router is started.  

## Related issue number

Closes https://github.com/ray-project/ray/issues/13180
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — APPROVED** [review]

**Comment by architkulkarni** [comment]
Tests pass on my laptop and on Mac CI, but on Linux CI `test_user_config_update` and `test_task_runner_perform_batch` fail.  Investigating...

File: python/ray/serve/tests/test_backend_worker.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def batcher(requests):
     query_param = make_request_param()
     my_batch_sizes = await asyncio.gather(*[(
         await router.assign_request.remote(query_param)) for _ in range(3)])
-    assert my_batch_sizes == [2, 2, 1]
+    assert sorted(my_batch_sizes) == [1, 2, 2]
```
Comment (simon-mo) [inline python/ray/serve/tests/test_backend_worker.py:229]: let's make sure to revert this and merge master again

**Review by edoakes — APPROVED** [review]

**Comment by simon-mo** [comment]
Hey @edoakes @architkulkarni this failed on Windows
image

```
==================== Test output for //python/ray/serve:test_handle:
============================= test session starts =============================
platform win32 -- Python 3.7.9, pytest-5.4.3, py-1.10.0, pluggy-0.13.1 -- C:\hostedtoolcache\windows\Python\3.7.9\x64\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\runneradmin\AppData\Local\Temp\Bazel.runfiles_kcsiur7i\runfiles\com_github_ray_project_ray
plugins: asyncio-0.14.0, rerunfailures-9.1.1, sugar-0.9.4, timeout-1.4.2
collecting ... collected 6 items

::test_async_handle_serializable 2021-01-27 20:56:30,837	INFO services.py:1182 -- View the Ray dashboard at http://127.0.0.1:8265
PASSED
::test_sync_handle_serializable PASSED
::test_handle_in_endpoint PASSED
::test_handle_http_args PASSED
::test_handle_inject_starlette_request PASSED
::test_handle_option_chaining PASSED

============================= 6 passed in 58.48s ==============================
(pid=6088) 2021-01-27 20:56:33,762	INFO http_state.py:70 -- Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:10.1.0.4-0' on node 'node:10.1.0.4-0' listening on '127.0.0.1:8000'

(pid=1660) INFO:     Started server process [1660]

(pid=6088) 2021-01-27 20:56:37,897	INFO controller.py:180 -- Registering route 'None' to endpoint 'f' with methods '['GET']'.

(pid=6088) 2021-01-27 20:56:39,072	INFO controller.py:190 -- Deleting endpoint 'f'

(pid=6952) 2021-01-27 20:56:39,056	INFO router.py:249 -- Endpoint f doesn't exist, waiting for registration.

(pid=6952) Windows fatal exception: access violation
(pid=6952) 
(pid=8036) Windows fatal exception: access violation
(pid=8036) 
(pid=6088) 2021-01-27 20:56:44,941	INFO controller.py:180 -- Registering route 'None' to endpoint 'f' with methods '['GET']'.

(pid=6088) 2021-01-27 20:56:46,127	INFO controller.py:190 -- Deleting endpoint 'f'

(pid=7680) 2021-01-27 20:56:46,111	INFO router.py:249 -- Endpoint f doesn't exist, waiting for registration.

(pid=1428) Windows fatal exception: access violation
(pid=1428) 
(pid=6088) 2021-01-27 20:56:50,969	INFO controller.py:180 -- Registering route '/endpoint1' to endpoint 'endpoint1' with methods '['GET', 'POST']'.

(pid=8144) 2021-01-27 20:56:54,140	WARNING api.py:512 -- You are retrieving a sync handle inside an asyncio loop. Try getting client.get_handle(.., sync=False) to get better performance. Learn more at https://docs.ray.io/en/master/serve/advanced.html#sync-and-async-handles

(pid=6088) 2021-01-27 20:56:56,013	INFO controller.py:180 -- Registering route '/endpoint2' to endpoint 'endpoint2' with methods '['GET', 'POST']'.

(pid=6088) 2021-01-27 20:56:58,065	INFO controller.py:190 -- Deleting endpoint 'endpoint1'

(pid=6088) 2021-01-27 20:56:58,068	INFO controller.py:190 -- Deleting endpoint 'endpoint2'

(pid=1660) 2021-01-27 20:56:58,047	INFO router.py:249 -- Endpoint endpoint2 doesn't exist, waiting for registration.

(pid=7680) Windows fatal exception: access violation
(pid=7680) 
(pid=8144) Windows fatal exception: access violation
(pid=8144) 
(pid=6088) 2021-01-27 20:57:05,068	INFO controller.py:180 -- Registering route '/endpoint' to endpoint 'endpoint' with methods '['POST']'.

(pid=6088) 2021-01-27 20:57:07,115	INFO controller.py:190 -- Deleting endpoint 'endpoint'

(pid=1172) Windows fatal exception: access violation
(pid=1172) 
(pid=6088) 2021-01-27 20:57:12,111	INFO controller.py:180 -- Registering route '/echo' to endpoint 'echo' with methods '['GET']'.

(pid=6088) 2021-01-27 20:57:17,162	INFO controller.py:180 -- Registering route '/wrapper' to endpoint 'wrapper' with methods '['GET']'.

(pid=6088) 2021-01-27 20:57:19,241	INFO controller.py:190 -- Deleting endpoint 'echo'

(pid=6088) 2021-01-27 20:57:19,244	INFO controller.py:190 -- Deleting endpoint 'wrapper'

(pid=5628) 2021-01-27 20:57:19,224	WARNING api.py:512 -- You are retrieving a sync handle inside an asyncio loop. Try getting client.get_handle(.., sync=False) to get better performance. Learn more at https://docs.ray.io/en/master/serve/advanced.html#sync-and-async-handles

(pid=5628) 2021-01-27 20:57:19,225	INFO router.py:249 -- Endpoint echo doesn't exist, waiting for registration.

(pid=8052) Windows fatal exception: access violation
(pid=8052) 
(pid=5628) Windows fatal exception: access violation
(pid=5628) 
(pid=6088) 2021-01-27 20:57:26,229	INFO controller.py:180 -- Registering route 'None' to endpoint 'm' with methods '['GET']'.

(pid=6088) 2021-01-27 20:57:26,248	INFO controller.py:190 -- Deleting endpoint 'm'

(pid=4908) Windows fatal exception: access violation
(pid=4908) 
================================================================================
(20:57:32) [419 / 497] 13 / 81 tests, 1 failed; Testing //python/ray/serve:test_kv_store; 3s local
(20:59:08) [423 / 501] 17 / 81 tests, 1 failed; Testing //python/ray/serve:test_router; 21s local
(20:59:23) FAIL: //python/ray/serve:test_router (see C:/users/runneradmin/_bazel_runneradmin/vlncal46/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_router/test_attempts/attempt_1.log)
```

https://github.com/ray-project/ray/runs/1779615703

So I'm reverting it now. Can you open a re-revert either fixing (seems unlikely, access violation could be port conflict but this seems hard to debug) or skip?

**Comment by architkulkarni** [comment]
So weird, thanks for catching this, and my fault for not checking the Windows CI.  It looks like this PR definitely caused this issue.  

I'm a little confused though, why does the run say that all 6/6 test_handle.py tests pass?  And why does it say "FAIL: //python/ray/serve:test_router" but there's no such line for test_handle?  Googling "Windows fatal exception: access violation" turns up a handful of pytest-related results, but also several that are not pytest-related.

The only idea I have for how to try to fix this is to use __getstate__ and __setstate__ as suggested by the Python docs instead of the lower-level __reduce__.  I don't know why this would work but I can give it a try.

**Comment by architkulkarni** [comment]
In the log https://github.com/ray-project/ray/runs/1779615703 I see a line above the handle tests saying "TIMEOUT: //python/ray/serve:test_handle", but I can't figure out which test is timing out since all 6/6 tests pass in each of the three runs.  The "Windows fatal exception" seems to be printed 1-2 times after each of the six tests, based on the endpoint names that are being printed.

**Pull request merged**